apps/chat/conversation.py

```diff
@@ -114,45 +115,75 @@ def compress_chat_history(
     input_messages: list,
     keep_history_len: int = 10,
 ) -> list[BaseMessage]:
+    history_messages = chat.get_langchain_messages_until_summary()
     try:
-        return _compress_chat_history(
-            chat,
+        history, last_message, summary = _compress_chat_history(
+            history=history_messages,
             llm=llm,
             max_token_limit=max_token_limit,
             input_messages=input_messages,
             keep_history_len=keep_history_len,
         )
+        if summary is not None:
+            ChatMessage.objects.filter(id=last_message.additional_kwargs["id"]).update(summary=summary)
+            return [SystemMessage(content=summary)] + history
+        return history
     except (NameError, ImportError, ValueError, NotImplementedError):
         # typically this is because a library required to count tokens isn't installed
         log.exception("Unable to compress history")
-        return chat.get_langchain_messages_until_summary()
+        return history_messages
 
 
-def _compress_chat_history(
-    chat: Chat, llm: BaseChatModel, max_token_limit: int, input_messages: list, keep_history_len: int = 10
+def compress_pipeline_chat_history(
```

Comment:
The only difference between this method and the `compress_chat_history` above it seems to be some class types i.e. `Chat` <-> `PipelineChatHistory` and `ChatMessage` <-> `PipelineChatMessages`. Looks like this method generally expects something that implements `get_langchain_messages_until_summary` and it needs some way of storing the new summary. I'm seeing some small abstraction that we can do here, maybe a thin wrapper class of sorts?

Reply:
Yes, you are correct, and it does feel like somewhat duplicated code, although I couldn't really think of another way of writing this that was as clear as this. I thought about adding the types as parameters to the original function, but it is less clear what is going on then.

Do you think adding a wrapper class would be clearer that what is happening here? Personally, I think that would just be a level of indirection more, but I'm happy to change it if you'd like me to.

Reply:
Nah if you've explored it and found that this implementation was clearer, then I'm happy. Thanks for clarifying :+1:

Reply:
Thanks!